### ISSUE TITLE
'emit' function not defined

### ISSUE THREAD

**Issue by VladyslavGoloshchapov** [pouchdb/pouchdb#4624]
hello.

I am creating secondary index like this
```javascript
                var assetSearchIndex = {
                    _id: '_design/assetSearchIndex',
                    views: {
                        byAssetNum: {
                            map: function(asset, emit) {
                                emit(asset.ASSETNUM);
                            }.toString()
                        }
                    }
                };
```
However when index is being built, I receive multiple errors
```
TypeError: emit is not a function
```
After some debugging I found out that I should not receive emit as param, because it exists in upper scope, when map function is eval-ed.

I think this is not correct, because:
1) it is inconsistent with temp index syntax
2) map is using 'emit' which is not passed nor defined (this also causes eslint validator errors)

Thanks,
Vlad

**Comment by fiatjaf** [comment]
You should declare your view like this:

```javascript
                var assetSearchIndex = {
                    _id: '_design/assetSearchIndex',
                    views: {
                        byAssetNum: {
                            map: (function(asset) {
                                emit(asset.ASSETNUM);
                            }).toString()
                        }
                    }
                };
```

You must believe `emit` will be defined at the time the function is run, and it will. This is the way CouchDB does it, and PouchDB mimics that for compatibility.

**Comment by VladyslavGoloshchapov** [comment]
sorry guys, but if it is possible to not use global variable, then one shouldn`t. this is one of the basic rules.
it is also bad idea to make "JSON object, toJSON, isArray, sum, etc" declared as global without any namespace, because those names are quite common and it will lead to conflicts with either other libraries or new browser API in future.

"Optional keyword-only arguments may help here, but I'm not sure if JS supports this feature or will in the nearest future."
this feature is not supported, but from my experience 2-3 arguments are not too much. In this case only 2 is enough.
 
@fiatjaf  - if you read my text above more attentively, you will see that I am aware that emit is defined on the run time (taken from upper scope).

**Comment by kxepal** [comment]
@VladyslavGoloshchapov Agreed, this is bad and not clear. Suggestions how to fix this and not break the BC? (: So far general solution leads to function signature for 7 arguments (for map function) with no oblivious order and way to handle this.

P.S. I spoke from CouchDB side, so there is no reason to care about any conflicts or browser API changes. Sure, if PouchDB follows the same, there could be some issues, but I'm not expert in this question (:

**Comment by akircher** [comment]
@kxepal you could slowly deprecate the globals (or BC in CouchDB 2.0), and meanwhile provide alternate syntax: `map:function(asset, globals)` where globals is defined as
```
{
   emit: function(){}
   toJSON: function() {}
   isArray: function() {}
...
}
```
Or maybe you don't need these at all as some "normal" JS globals already contain many of these features e.g., Array.isArray(), JSON.stringify(), etc

**Comment by VladyslavGoloshchapov** [comment]
@kxepal, sorry I am a bit new to Pouch/Couch. 

As far as I know only 2 params are required for map function: (doc (object), emit(function)). did I miss something?

again if there are really 7 params - it is possible to combine them into one. JS does not support names params, but that can be easily substituted with object:

```
fun.apply(null, {doc: doc, emit: emit, smthElse: smthElse ...})

```
I completely understand that this may be a big change to PouchDB, but having many globals is worse in my opinion.
usually libraries export only 1 global variable, like $. but even jquery has .noConflict() function.

@akircher completely agree.

**Comment by kxepal** [comment]
@VladyslavGoloshchapov `emit` is just a vital function for maps. Basically on CouchDB you have the following picture: http://docs.couchdb.org/en/1.6.1/query-server/javascript.html

So, if I understand, the only solution here is to wrap all these functions into object and pass them explicitly? Like
```
function(doc, ctx){
  if (doc.tags && ctx.isArray(doc.tags)){
    ctx.emit([doc._id, null]);
  }
}
```

Right? This could be a good idea, but for JS it's easy to forget specify this argument and get lost. However, it's newbie mistake and would be easy to correct yourself on the second iteration.

Hiding all current injected functions into global namespace behind some magic $ makes no sense. The main profit from functions-as-an-argument for me is that you replace these functions with your own for unit testing, for example. Second is yours - explicit things makes linter happy.

**Comment by VladyslavGoloshchapov** [comment]
@kxepal yes - this is good idea.

"Hiding all current injected functions into global namespace behind some magic $ makes no sense. The main profit from functions-as-an-argument for me is that you replace these functions with your own for unit testing, for example. Second is yours - explicit things makes linter happy."

agree. I just noted this as an example of what others are doing.

p.s. looked into http://docs.couchdb.org/en/1.6.1/query-server/javascript.html
in my opinion Pouch is supplying too many helpers. The following helpers are not responsibility of Pouch itself, but responsibility of a user of Pouch:
* JSON (this is already standard object. needs polyfill only for very old browsers)
* isArray()
* toJSON()
* require()
Anyone can easily add those if needed. In most cases, I think, people don`t need them at all or can use standard browser API.

If you remove them, Pouch will be lighter and will not make everyone use libraries they don`t need.

**Comment by kxepal** [comment]
@VladyslavGoloshchapov Ok, thank you. I'll throw this idea to CouchDB ML and will see how it goes. Leaving this thread for PouchDB discussion now (:

**Comment by VladyslavGoloshchapov** [comment]
thank you for attention

**Comment by nolanlawson** [comment]
@VladyslavGoloshchapov Yeah, we inherited a lot of bad practices from CouchDB. Keep in mind that CouchDB is almost 10 years old, and the number 1 priority of PouchDB is to imitate it exactly.

If you dislike globals this much, maybe look into pouchdb-find? :smiley: https://github.com/nolanlawson/pouchdb-find

**Comment by kxepal** [comment]
I return feedback from CouchDB here:
https://pony-poc.apache.org/thread.html/Zi5vpflalivhdg2

TL:DR; We all agree to change this. Most popular solution is to use `this.` approach to not uglify function signature. Most possible candidate to get these changes is 3.0 release. Thanks for raising this question!

**Comment by VladyslavGoloshchapov** [comment]
@kxepal thank you too

**Comment by alexweber** [comment]
@VladyslavGoloshchapov for what it's worth, I stumbled upon this same issue using Pouch with Ionic 2 and was able to work around it by trusting @nolanlawson that emit is in the global scope and cheating the ts compiler with `declare function emit (val: any);`.  (not tested extensively just yet but seems to be ok)

**Comment by VladyslavGoloshchapov** [comment]
I also am using that workaround, but the problem is that I don`t like hacks.

global variables with such common name should not exist.

**Comment by thiagobackadelino** [comment]
obtive o mesmo problema utilizando typescript e pouchdb , couchdb ao criar Persistent queries.
como solução e levando em conta que não será o type script que executara a function e sim o couchdb
 

![](https://cloud.githubusercontent.com/assets/14223502/22002217/dd0823a0-dc28-11e6-89a1-4284dd319c35.png)

**Comment by seanjennings960** [comment]
Just seeing this now old thread... A thought for others who might stumble upon this.

>for what it's worth, I stumbled upon this same issue using Pouch with Ionic 2 and was able to work around it by trusting @nolanlawson that emit is in the global scope and cheating the ts compiler with declare function emit (val: any);.

I actually find this an reasonably acceptable solution. Let me explain my reasoning. Since this is a "design document" that presumably could be being executed in some other context (such a remote CouchDB server, although I'm not sure this is necessarily the case with all PouchDB adapters). Notice the `.toString` at the end of the line:
```javascript
let ddoc = {
    _id: '_design/assetSearchIndex',
    views: {
        byAssetNum: {
            map: (function(asset) {
                emit(asset.ASSETNUM);
            }).toString()
        }
    }
)
```
What's being saved in `ddoc['views'][viewName]` is just a string. I think of `emit` in this context as a *keyword* in a CouchDB-specific DSL (domain-specific language). The fact that it's written as a Javascript function before serialization is works because this DSL is also based on Javscript. But this is merely a convenience that allows us to format the code a bit nicer.  @thiagobackadelino 's reformatting of `emit` into a string makes this blatantly obvious. Although I'd call the variable something more along the lines of `mapFunc`, his comment lead me to this thought:
```javascript
let mapFunc = "function(doc) { if (doc.type === "example" { emit(doc.name) }"
```
You could convert this statement to a Python function, if for instance you wanted to implement a Python version of CouchDB. Not that any of this would necessarily make sense to do, but it hopefully you get the point.

`emit` is a function with very specific sematics within the CouchDB map-reduce framework. Using `declare` to take advantage of Syntax highlighting and TypeScript checking, seems like a reasonable solution to me.

```
I return feedback from CouchDB here:
https://pony-poc.apache.org/thread.html/Zi5vpflalivhdg2

TL:DR; We all agree to change this. Most popular solution is to use this. approach to not uglify function signature. Most possible candidate to get these changes is 3.0 release. Thanks for raising this question!
```
(Of course, not creating a DSL and using the `this` keyword also sounds like an even better idea. But not all APIs are perfect and sometimes we don't want to wait for the next release ;-)
